composer/trainer/deepspeed.py

```diff
@@ -1,98 +1,121 @@
+        raise ValueError("The DeepSpeed configuration specifies an optimizer, but the Mosaic "
+                         "trainer will override this setting.")
+
+    if "scheduler" in config:
+        raise ValueError("The DeepSpeed configuration specifies a scheduler, but the Mosaic "
+                         "trainer will override this setting.")
+
+
+def parse_precision_settings(config: Dict[str, Any], state: State):
+    precision = state.precision
+
+    ds_precision = None
+    if "fp16" in config and "enabled" in config["fp16"] and config["fp16"]["enabled"]:
+        ds_precision = Precision.FP16
+    if "bf16" in config and "enabled" in config["bf16"] and config["bf16"]["enabled"]:
+        raise ValueError("DeepSpeed is configured to use BFLOAT16, but this is unsupported by the "
+                         "Mosaic trainer.")
+    if "amp" in config and "enabled" in config["amp"] and config["amp"]["enabled"]:
+        raise ValueError("DeepSpeed is configured to use Apex AMP, but this is unsupported by the "
+                         "Mosaic trainer.")
+
+    if ds_precision is not None and ds_precision != precision:
+        raise ValueError(f"Provided DeepSpeed configuration specifies precision={ds_precision}, "
+                         f"but the Mosaic trainer has been configured with precision={precision}.")
+
+    if precision == Precision.FP16:
+        if "fp16" not in config:
+            config["fp16"] = cast({"enabled": True}, Dict[str, Any])
+        fp16_config = config["fp16"]
+        assert isinstance(fp16_config, dict)
+
+        # For equivalence with the non-DeepSpeed defaults of the Mosaic trainer.
+        fp16_config.setdefault("initial_scale_power", 16)
+        fp16_config.setdefault("loss_scale_window", 2000)
+
+
+def parse_misc_settings(config: Dict[str, Any], grad_clip_norm: Optional[float]):
+    if "gradient_clipping" in config:
+        ds_grad_clip_norm = config["gradient_clipping"]
+        if ds_grad_clip_norm != grad_clip_norm:
+            raise ValueError("Provided DeepSpeed configuration specifies grad clip norm="
+                             f"{ds_grad_clip_norm}, but the Mosaic trainer has been configured "
+                             f"with grad clip norm={grad_clip_norm}")
+
+    if grad_clip_norm is not None:
+        config["gradient_clipping"] = grad_clip_norm
+
+    if "zero_allow_untested_optimizer" in config and not config["zero_allow_untested_optimizer"]:
+        warnings.warn("Provided DeepSpeed configuration specifies zero_allow_untested_optimizer=False. "
+                      "This causes DeepSpeed to reject certain Mosaic optimizers that are known to "
+                      "work well with DeepSpeed.")
+
+    config["zero_allow_untested_optimizer"] = True
+
+
+def parse_deepspeed_config(config: Dict[str, Any],
+                           state: State,
+                           grad_clip_norm: Optional[float] = None) -> Dict[str, Any]:
+    new_config = copy.deepcopy(config)
```

Comment:
1. Docstring
2. Would it make sense to modify the config in place rather than deep-copying?

Reply:
The config is potentially user provided, so I'd rather not modify it if possible.

I guess we violate this principle for other things used to initialize the trainer, like the model and optimizers, but any reasonable person would _expect_ those to be modified. I'm less sure about something like a DeepSpeed config.